src/bluetooth_linux.c

```diff
@@ -0,0 +1,466 @@
+		if (ret >= 0) {
+			if (!strcmp(propname, "Connected") && boolval)
+				connected = true;
+			if (!strcmp(propname, "Paired") && boolval)
+				paired = true;
+			if (!strcmp(propname, "ServicesResolved") && boolval)
+				resolved = true;
+		} else {
+			sd_bus_message_rewind(reply, 0);
+			ret = sd_bus_message_read_basic(reply, 's', &propname);
+			if (ret >= 0 && !strcmp(propname, "Name") &&
+			    name != NULL && sd_bus_message_read(reply, "v", "s", name) >= 0) {}
+			if (ret >= 0 && !strcmp(propname, "UUIDs") &&
+			   (sd_bus_message_enter_container(reply, SD_BUS_TYPE_VARIANT, "as") >= 0)) {
+				if (sd_bus_message_enter_container(reply, SD_BUS_TYPE_ARRAY, "s") >= 0) {
+					const char *uuid;
+					while(sd_bus_message_read_basic(reply, 's', &uuid)) {
+						if (!strcasecmp(uuid, FIDO_SERVICE_UUID))
+							has_service = true;
+					}
+					sd_bus_message_exit_container(reply); /* s */
+				}
+				sd_bus_message_exit_container(reply); /* as */
+			} else {
+				sd_bus_message_skip(reply,"v");
+			}
+		}
+		sd_bus_message_exit_container(reply); /* sv */
+	}
+	sd_bus_message_exit_container(reply);  /* {sv} */
+	return (allow_unconnected || connected) && (allow_unconnected || resolved) && has_service && paired;
+}
+
+static int init_ble_fido_dev(fido_dev_info_t *di,
+			      const char *path, const char *name)
+{
+	memset(di, 0, sizeof(*di));
+	if (asprintf(&di->path, "%s%s", FIDO_BLUETOOTH_PREFIX, path) &&
+		(di->manufacturer = strdup("BLE")) &&
+		(di->product = strdup(name))) {
+		di->io = (fido_dev_io_t) {
+			fido_bluetooth_open,
+			fido_bluetooth_close,
+			fido_bluetooth_read,
+			fido_bluetooth_write,
+		};
+		di->transport = (fido_dev_transport_t) {
+			fido_bluetooth_rx,
+			fido_bluetooth_tx,
+		};
+
+		return 0;
+	}
+
+	free(di->product);
+	free(di->manufacturer);
+	free(di->path);
+	explicit_bzero(di, sizeof(*di));
+
+	return -1;
+}
+
+static void fido_bluetooth_add_device(void *data, const char *path, sd_bus_message *reply)
+{
+	struct manifest_ctx *ctx = (struct manifest_ctx *) data;
+	const char *iface;
+	if (sd_bus_message_read_basic(reply, 's', &iface) > 0) {
+		const char *name;
+		if (ble_fido_is_useable_device(iface, reply, false, &name)) {
+			if (!init_ble_fido_dev(&ctx->devlist[*ctx->olen], path, name)) {
+				if (++(*ctx->olen) == ctx->ilen)
+					return;
+			}
+		}
+		sd_bus_message_rewind(reply, 0);
+		sd_bus_message_skip(reply, "sa{sv}");
+	}
+}
+
+int
+fido_bluetooth_manifest(fido_dev_info_t *devlist, size_t ilen, size_t *olen)
+{
+	sd_bus *bus;
+	sd_bus_message *reply;
+	int ret;
+	struct manifest_ctx ctx;
+
+	*olen = 0;
+	if (ilen == 0)
+		return FIDO_OK;
+	if (devlist == NULL)
+		return FIDO_ERR_INVALID_ARGUMENT;
+
+	ctx.devlist = devlist;
+	ctx.olen = olen;
+	ctx.ilen = ilen;
+	if (0>sd_bus_default_system(&bus))
+		return FIDO_ERR_INTERNAL;
+
+	ctx.bus = bus;
+	ret = sd_bus_call_method(bus, "org.bluez", "/", "org.freedesktop.DBus.ObjectManager",
+				 "GetManagedObjects", NULL, &reply, "");
```

Comment:
Nit: For second level indentation, we use four spaces.

```suggestion
	ret = sd_bus_call_method(bus, "org.bluez", "/", "org.freedesktop.DBus.ObjectManager",
	    "GetManagedObjects", NULL, &reply, "");
```